### PR TITLE
Versioning and locking documentation

### PR DESCRIPTION
**Pull request by bmichalski**
Works on fixing #1585.

### CODE REVIEW

File: docs/en/reference/annotations-reference.rst [hunk 1]
```diff
@@ -1412,6 +1412,20 @@ encouraged to use the :ref:`atomicSet <atomic_set>` or
 will ensure that collections are updated in the same write operation as the
 versioned document.
 
+@Lock
```
Comment (alcaeus) [inline docs/en/reference/annotations-reference.rst]: The annotation documentation is sorted alphabetically, could you move this between `@Key` and `@MappedSuperclass`?

File: docs/en/reference/annotations-reference.rst [hunk 2]
```diff
@@ -1394,8 +1394,8 @@ Alias of `@Index`_, with the ``unique`` option set by default.
 --------
 
 The annotated instance variable will be used to store version information, which
-is used for pessimistic and optimistic locking. This is only compatible with
-integer and date field types, and cannot be combined with `@Id`_.
+is used for optimistic locking. This is only compatible with
```
Comment (jmikola) [inline docs/en/reference/annotations-reference.rst]: Should "optimistic locking" link to a header in https://github.com/doctrine/mongodb-odm/pull/1587?

File: docs/en/reference/annotations-reference.rst [hunk 3]
```diff
@@ -1412,6 +1412,20 @@ encouraged to use the :ref:`atomicSet <atomic_set>` or
 will ensure that collections are updated in the same write operation as the
 versioned document.
 
+@Lock
+-----
+
+The annotated instance variable will be used to store lock information, which
+is used for pessimistic locking. This is only compatible with
```
Comment (jmikola) [inline docs/en/reference/annotations-reference.rst]: Should "pessimistic locking" link to a header in https://github.com/doctrine/mongodb-odm/pull/1587?

File: docs/en/reference/xml-mapping.rst [hunk 4]
```diff
@@ -147,3 +147,40 @@ Be aware that class-names specified in the XML files should be fully qualified.
...
     ``field-name`` if not set explicitly.
+
+Reference
+---------
+
+Version
+^^^^^^^
+
+The field with the ``version`` attribute will be used to store version information, which
+is used for optimistic locking. This is only compatible with
```
Comment (jmikola) [inline docs/en/reference/xml-mapping.rst]: Should "optimistic locking" link to a header in https://github.com/doctrine/mongodb-odm/pull/1587?

File: docs/en/reference/xml-mapping.rst [hunk 5]
```diff
@@ -147,3 +147,40 @@ Be aware that class-names specified in the XML files should be fully qualified.
...
+``int`` and ``date`` field types, and cannot be combined with ``id="true"``.
+
+.. code-block:: xml
+
+    <doctrine-mongo-mapping>
+        <field fieldName="version" version="true" type="int" />
+    </doctrine-mongo-mapping>
+
+By default, Doctrine ODM processes updates :ref:`embed-many <embed_many>` and
```
Comment (jmikola) [inline docs/en/reference/xml-mapping.rst]: Remove "processes", as "updates" already works as a verb here.

File: docs/en/reference/xml-mapping.rst [hunk 6]
```diff
@@ -147,3 +147,40 @@ Be aware that class-names specified in the XML files should be fully qualified.
...
+:ref:`reference-many <reference_many>` collections in separate write operations,
+which do not bump the document version. Users employing document versioning are
+encouraged to use the :ref:`atomicSet <atomic_set>` or
+:ref:`atomicSetArray <atomic_set_array>` strategies for such collections, which
+will ensure that collections are updated in the same write operation as the
+versioned document.
```
Comment (jmikola) [inline docs/en/reference/xml-mapping.rst]: "versioned parent document".

@alcaeus: I assume versioning is limited to root documents. Do we actually mention this anywhere? I don't recall seeing it in https://github.com/doctrine/mongodb-odm/pull/1587.

File: docs/en/reference/xml-mapping.rst [hunk 7]
```diff
@@ -147,3 +147,40 @@ Be aware that class-names specified in the XML files should be fully qualified.
...
+
+Lock
+^^^^
+
+The field with the ``lock`` attribute will be used to store lock information, which
+is used for pessimistic locking. This is only compatible with
```
Comment (jmikola) [inline docs/en/reference/xml-mapping.rst]: "used to store lock information for pessimistic locking". Also considering linking "pessimistic locking" to a header in https://github.com/doctrine/mongodb-odm/pull/1587.

File: docs/en/reference/xml-mapping.rst [hunk 8]
```diff
@@ -147,3 +147,40 @@ Be aware that class-names specified in the XML files should be fully qualified.
...
+``int`` field type, and cannot be combined with ``id="true"``.
```
Comment (jmikola) [inline docs/en/reference/xml-mapping.rst]: Add "the":

> This is only compatible with the `int` field type.

File: docs/en/reference/yml-mapping.rst [hunk 9]
```diff
@@ -170,3 +170,40 @@ Be aware that class-names specified in the YAML files should be fully qualified.
...
     of mapped field.
+
+Reference
+---------
+
+Version
+^^^^^^^
+
+The field with the ``version`` property will be used to store version information, which
+is used for optimistic locking. This is only compatible with
```
Comment (jmikola) [inline docs/en/reference/yml-mapping.rst]: "used to store version information for optimistic locking". Also consider linking "optimistic locking" to a header in #1587.

File: docs/en/reference/yml-mapping.rst [hunk 10]
```diff
@@ -170,3 +170,40 @@ Be aware that class-names specified in the YAML files should be fully qualified.
...
+``int`` and ``date`` field types, and cannot be combined with ``id: true``.
+
+.. code-block:: yaml
+
+    version:
+      type: int
+      version: true
+
+By default, Doctrine ODM processes updates :ref:`embed-many <embed_many>` and
```
Comment (jmikola) [inline docs/en/reference/yml-mapping.rst]: Remove "processes", as "updates" already works as a verb here.

File: docs/en/reference/yml-mapping.rst [hunk 11]
```diff
@@ -170,3 +170,40 @@ Be aware that class-names specified in the YAML files should be fully qualified.
...
+:ref:`reference-many <reference_many>` collections in separate write operations,
+which do not bump the document version. Users employing document versioning are
+encouraged to use the :ref:`atomicSet <atomic_set>` or
+:ref:`atomicSetArray <atomic_set_array>` strategies for such collections, which
+will ensure that collections are updated in the same write operation as the
+versioned document.
```
Comment (jmikola) [inline docs/en/reference/yml-mapping.rst]: "versioned parent document".

File: docs/en/reference/yml-mapping.rst [hunk 12]
```diff
@@ -170,3 +170,40 @@ Be aware that class-names specified in the YAML files should be fully qualified.
...
+
+Lock
+^^^^
+
+The field with the ``lock`` property will be used to store lock information, which
+is used for pessimistic locking. This is only compatible with
+``int`` field type, and cannot be combined with ``id: true``.
```
Comment (jmikola) [inline docs/en/reference/yml-mapping.rst]: "the `int` field type"

File: docs/en/reference/yml-mapping.rst [hunk 13]
```diff
@@ -170,3 +170,40 @@ Be aware that class-names specified in the YAML files should be fully qualified.
...
+encouraged to use the :ref:`atomicSet <atomic_set>` or
+:ref:`atomicSetArray <atomic_set_array>` strategies for such collections, which
+will ensure that collections are updated in the same write operation as the
+versioned document.
+
+Lock
+^^^^
+
+The field with the ``lock`` property will be used to store lock information, which
+is used for pessimistic locking. This is only compatible with
```
Comment (jmikola) [inline docs/en/reference/yml-mapping.rst]: "used to store lock information for pessimistic locking". Also considering linking "pessimistic locking" to a header in https://github.com/doctrine/mongodb-odm/pull/1587.

File: docs/en/reference/xml-mapping.rst (hunk 6 above)
Comment (alcaeus) [inline docs/en/reference/xml-mapping.rst]: Yes, versioning is limited to root documents. IIRC this was one of the reasons why @malarzm added the atomic strategies.

Comment (alcaeus) [inline docs/en/reference/xml-mapping.rst]: Oh, and no, we don't mention that anywhere. @bmichalski could you add a note in #1587 for this?

File: docs/en/reference/annotations-reference.rst (hunk 2 above)
Comment (bmichalski) [inline docs/en/reference/annotations-reference.rst]: That is a good idea. I will have to create unique section labels and reference them with a link from another document (as in http://www.sphinx-doc.org/en/stable/markup/inline.html).

File: docs/en/reference/xml-mapping.rst (hunk 6 above)
Comment (bmichalski) [inline docs/en/reference/xml-mapping.rst]: I will :)

**Comment by bmichalski** [comment]
All requested changes have been integrated :)

**Comment by malarzm** [comment]
I see that @alcaeus' remark has been applied and with @jmikola's seal of approval I'm more than happy to merge this - thanks a lot @bmichalski!

**Pull request merged**